database/interface.py

```diff
@@ -142,13 +148,102 @@ def update_harvest_source(self, source_id, updates):
             self.db.rollback()
             return None
 
+    def clear_harvest_source(self, source_id):
+        """
+        Clear all datasets related to a harvest source in CKAN, and clean up the
+        harvest_record and harvest_record_error tables.
+        :param source_id: ID of the harvest source to clear
+        """
+
+        def _clear_harvest_records():
+                self.db.query(HarvestRecordError).filter(
+                    HarvestRecordError.harvest_record_id.in_(
+                        self.db.query(HarvestRecord.id).filter_by(harvest_source_id=source_id)
+                    )
+                ).delete(synchronize_session=False)
+                self.db.query(HarvestRecord).filter_by(harvest_source_id=source_id).delete()
+                self.db.commit()
+
+        source = self.db.get(HarvestSource, source_id)
+        if source is None:
+            return "Harvest source not found"
+
+        organization_id = source.organization_id
+
+        records = (
+            self.db.query(HarvestRecord)
+            .filter_by(harvest_source_id=source_id).all()
+        )
+        if not records:
+            return "Harvest source has no records to clear."
+
+        ckan_ids = [record.ckan_id for record in records if record.ckan_id is not None]
+
+        if not ckan_ids:  
+            _clear_harvest_records()
+            return "Harvest source cleared successfully."
+
+        ckan = RemoteCKAN(
+            os.getenv("CKAN_API_URL"), apikey=os.getenv("CKAN_API_TOKEN")
+        )
+
+        retry_count = 0
+        result = ckan.action.package_search(fq=f'owner_org:{organization_id}')
+        ckan_datasets = result['count']
+        start = datetime.now(timezone.utc)
+        logger.info(f"Attempting to delete {ckan_datasets} datasets at {start}")
+
+        while ckan_datasets > 0 and retry_count < 20:
+            try:
+                ckan.action.bulk_update_delete(
+                    datasets=ckan_ids, org_id=organization_id
+                )
+                result = ckan.action.package_search(fq=f'owner_org:{organization_id}')
+                ckan_datasets = result['count']
+
+                if ckan_datasets == 0:
+                    _clear_harvest_records()
+                    logger.info(f"Total time: {datetime.now(timezone.utc) - start}")
+                    return "Harvest source cleared successfully."
+                else:
+                    logger.info(f"Retry {retry_count + 1}: "
+                                "{ckan_datasets} datasets remaining, retrying...")
```

Comment:
These lines do not belong to this try block.

Reply:
right, it runs only after clean up all dataset which outside of the try block.

Reply:
fixed.